client/sftp_client.py

```diff
@@ -196,202 +614,749 @@ def load_dirs(self, path, parent=None):
+		self.remote_dropdown.set_entry_text_column(0)
+		self.remote_dropdown.connect('changed', self.signal_combo_changed, self.remote)
+		box_remote.pack_start(self.remote_dropdown, True, True, 0)
+
+		self.window.connect('destroy', self.signal_window_destroy)
 		self.window.show_all()
 
-	def upload(self, temp, temp1):
-		selection = self.local.treeview_local.get_selection()
+
+	def signal_combo_changed(self, combo, *data):
+		system = data[0]
+		treeiter = combo.get_active_iter()
+		model = combo.get_model()
+		if treeiter != None:
+			new_dir = model[treeiter][0]
+		else:
+			# the user has typed something into the combo box
+			new_dir = '.'
+			entry = combo.get_child().get_text()
+			if entry == '..':
+				pass
+			elif system._already_exists_all(entry) and entry != system.default_directory:
+				if system.get_is_folder(entry) and 'r' in system._check_perm(entry):
+					# if the entered string is valid
+					new_dir = entry
+		if new_dir != '.':
+			if new_dir == '..':
+				split = system.default_directory.split('/')
+				split = split[:-1]
+				new_dir = '/'.join(split)
+				if new_dir == '':
+					new_dir = '/'
+			system.default_directory = new_dir
+			system._tv_model.clear()
+			system.load_dirs(system.default_directory)
+			model.clear()
+			self.render_dropdown(model, system)
+
+	def render_dropdown(self, model, system):
+		model.append(('..', ))
+		for _dir in system._yield_dir_list(system.default_directory):
+			filename = os.path.join(system.default_directory, _dir)
+			if 'r' not in system._check_perm(filename):
+				continue
+			elif not system.get_is_folder(filename):
+				continue
+			elif system.local_hidden and _dir.startswith('.'):
+				continue
+			model.append((filename,))
+
+	def render_menubar(self, menubar):
+		self.validate = False
+		menu_item = Gtk.CheckMenuItem.new_with_label('Checksum Validation')
+		menu_item.connect('toggled', self.signal_toggled_validate_checksums)
+		self.validate_checksums = menu_item
+		menubar.append(menu_item)
+
+		self.transfer_hidden = True
+		menu_item = Gtk.CheckMenuItem.new_with_label('Transfer Hidden Files')
+		menu_item.connect('toggled', self.signal_toggled_transfer_hidden)
+		self.signal_toggled_transfer_hidden = menu_item
+		menubar.append(menu_item)
+
+		menu_item = Gtk.SeparatorMenuItem()
+		menubar.append(menu_item)
+
+		menu_item = Gtk.MenuItem.new_with_label('Exit')
+		menu_item.connect('activate', self.shutdown)
+		menubar.append(menu_item)
+
+	def shutdown(self, _):
+		self.signal_window_destroy(None)
+		self.window.destroy()
+
+	def signal_toggled_transfer_hidden(self, _):
+		self.transfer_hidden = not self.transfer_hidden
+
+	def signal_toggled_validate_checksums(self, _):
+		self.validate = not self.validate
+
+	def _transfer_folder(self, task, ssh):
+		task.state = 'Transferring'
+		if task.parents is not None:
+			# prevents any lagging folders/files from causing issues
+			if  task.parents[0].state == 'Cancelled':
```

Comment:
There are two spaces after the if here and only one is necessary, additionally this and the code below it should be combined like:
```
if task.parents[0].state in ('Cancelled', 'Paused'):
    task.state = task.parents[0].state
    return
```